crates/matrix-sdk-ui/src/timeline/inner/mod.rs

```diff
@@ -141,26 +145,46 @@ impl<P: RoomDataProvider> TimelineInner<P> {
     ///
     /// Cheap because `im::Vector` is cheap to clone.
     pub(super) async fn items(&self) -> Vector<Arc<TimelineItem>> {
-        self.state.lock().await.items.clone()
+        self.state.read().await.items.clone()
     }
 
-    pub(super) async fn subscribe(
-        &self,
-    ) -> (Vector<Arc<TimelineItem>>, VectorSubscriber<Arc<TimelineItem>>) {
+    pub(super) fn subscribe(&self) -> impl Stream<Item = VectorDiff<Arc<TimelineItem>>> {
         trace!("Creating timeline items signal");
-        let state = self.state.lock().await;
-        // auto-deref to the inner vector's clone method
-        let items = state.items.clone();
-        let stream = state.items.subscribe();
-        (items, stream)
+        let state = self.state.clone();
+        let ignore_user_list_stream = self.client.subscribe_to_ignore_user_list_changes();
+
+        stream! {
+            pin_mut!(ignore_user_list_stream);
+
+            loop {
+                let (timeline_items, timeline_stream) = {
+                    let state = state.read().await;
+
+                    (state.items.clone(), state.items.subscribe())
+                };
+
+                yield stream::once(
+                    // Reset the stream with all its items.
+                    ready(VectorDiff::Reset { values: timeline_items } ),
+                )
+                .chain(
+                    timeline_stream
+                );
+
+                // When this is fired, let's loop.
+                let _ = ignore_user_list_stream.next().await;
+
+                // A user has been ignored/unignored? Let's clear the timeline.
+                state.write().await.clear();
+            }
+        }
+        .switch()
```

Comment:
I don't think this should be part of the "inner" bits of the timeline. Can you revert all the changes in this file and the unit tests please, and move the listening for the ignoreu user list changes one level higher, in the `timeline` module? (also, I think the `RwLock` thing should be a separate PR, if we want to do it at all – if we do the `eyeball-im` changes discussed today we probably don't need it)